### PR TITLE
ramfs:Fixing multiple spaces between commands and merging cmds and extra flag into just cmds

### PR DESCRIPTION
**Pull request by ananion-cation**
1. Allows for multiple spaces ex: "a    (4 spaces) b c d" will  be interpreted as [a,b,c,d]
2.Combine cmd and extra flag so it is just cmd

### CODE REVIEW

File: bb/ramfs.go [hunk 1]
```diff
@@ -115,23 +117,57 @@ func ramfs() {
...
+				copyCommands(w, eachPath)
 			}
-			paths[p[0]] = append(paths[p[0]], p[1])
 		}
 	}
 
-	if *extraCmds != "" {
-		copyCommands(w, strings.Split(*extraCmds, " "))
-	}
+	/*
```
Comment (rminnich) [inline bb/ramfs.go]: this is all commented out?

Comment (ananion-cation) [inline bb/ramfs.go]: They are commented out now because both extra paths and extra commands have been consolidated - should I remove this comment block?

Comment (hugelgupf) [inline bb/ramfs.go]: Don't keep dead code around. Just remove it. We can use git history to see what used to be there if we need to.

File: bb/ramfs.go [hunk 2]
```diff
@@ -1,5 +1,5 @@
 // Copyright 2015-2017 the u-root Authors. All rights reserved
-// Use of this source code is governed by a BSD-style
+// Use of this source code is governed by a BSiD-style
```
Comment (rminnich) [inline bb/ramfs.go]: BSD, not BSiD

File: bb/ramfs.go [hunk 3]
```diff
@@ -70,15 +67,17 @@ func dirComponents(dir string) []string {
 
 // copyCommands takes a list of commands, generates the list of libs,
 // and creates cpio records, including directory records.
-func copyCommands(w cpio.Writer, cmds []string) {
-	debug("copyCommands: start with %v", cmds)
+func copyCommands(w cpio.Writer, cmd string) {
+	debug("copyCommands: start with %v", cmd)
 	var recs []cpio.Record
-	libs, err := uroot.LddList(cmds)
+	//introduced the tmpSlice variable to be the dependency slice for the cmd string
```
Comment (rminnich) [inline bb/ramfs.go]: You don't need this comment I think, or at least the word introduced.

File: bb/ramfs.go [hunk 4]
```diff
@@ -115,23 +114,29 @@ func ramfs() {
...
-	if *extraPaths != "" {
-		extras := strings.Split(strings.TrimSpace(*extraPaths), " ")
-		for _, x := range extras {
-			p := strings.Split(x, ":")
-			if len(p) != 2 {
-				p = append([]string{"/"}, p...)
+	if *extraCmds != "" {
+		copyc := strings.Fields(*extraCmds)
+		for _, eachPath := range copyc {
+			// eachPath must be an absolute filepath for os.Stat
```
Comment (rminnich) [inline bb/ramfs.go]: That's not really right. It can be  a path relative to cwd. I think you can remove this comment.

File: bb/ramfs.go [hunk 5]
```diff
@@ -115,23 +114,29 @@ func ramfs() {
...
+			statval, err := os.Stat(eachPath)
```
Comment (rminnich) [inline bb/ramfs.go]: Did you test this command with a string containing a :? Below, you split eachPath on the :. But here, you stat the path with no change. If the path contains a :, then the stat will fail. I think a change this far reaching might need to include a test.

Comment (ananion-cation) [inline bb/ramfs.go]: That's true - I'll add the test case and change the code accordingly

File: bb/ramfs.go [hunk 6]
```diff
@@ -115,23 +113,35 @@ func ramfs() {
...
-			if len(p) != 2 {
-				p = append([]string{"/"}, p...)
+	if *extraCmds != "" {
+		copyc := strings.Fields(*extraCmds)
+		//check if the path is a file or directory
+		for _, eachPath := range copyc {
+			//must not include ~ in path
+			//replace only the first : with //
+			modPath := strings.Replace(eachPath, ":", "", 1)
+			fmt.Println("this is the path", modPath)
```
Comment (LaconicPneumonic) [inline bb/ramfs.go]: Consider changing the Println to a debug

File: bb/ramfs.go [hunk 7]
```diff
@@ -28,10 +28,7 @@ var (
 	// e.g., /lib/modules/4.04, you would do add the root as / and the
 	// starting point for the walk as lib/modules/4.04. That way we only preserve
 	// as much of the path as we need, but we can preserve it all.
-	paths      = map[string][]string{}
-	extraPaths = flag.String("extra", "", `Extra paths to add in the form root:start, e.g. /:etc/hosts.
-The path before the : is used as a starting point for a walk; the path after the : selects what things to put
-into the initramfs starting at /. E.g., /tmp/prototype:/ will install the prototype file system into / of the initramfs`)
+	paths = map[string][]string{}
```
Comment (hugelgupf) [inline bb/ramfs.go]: It's sufficient to write `map[string][]string` (tbh, I had no idea you could put the `{}` at the end and it threw me off)

File: bb/ramfs.go [hunk 8]
```diff
@@ -115,23 +113,35 @@ func ramfs() {
...
+		copyc := strings.Fields(*extraCmds)
+		//check if the path is a file or directory
+		for _, eachPath := range copyc {
+			//must not include ~ in path
+			//replace only the first : with //
+			modPath := strings.Replace(eachPath, ":", "", 1)
+			fmt.Println("this is the path", modPath)
+			statval, err := os.Stat(modPath)
+			if err != nil {
+				log.Fatalf("%v\n", err)
```
Comment (hugelgupf) [inline bb/ramfs.go]: No need to end with newline in log, log will append it automatically

Comment (hugelgupf) [inline bb/ramfs.go]: We could convert all of this to returning errors, if you feel like doing that

File: bb/ramfs.go (hunk 7 above)
Comment (ananion-cation) [inline bb/ramfs.go]: ./ramfs.go:31:10: type map[string][]string is not an expression

Comment (rminnich) [inline bb/ramfs.go]: you have to put the {} at the end. It creates an empty map.

Comment (rminnich) [inline bb/ramfs.go]: otherwise you would need to say paths=make(...) to make it non nil

**Comment by rminnich** [comment]
if you strace you'll see it stats /etc//hosts properly, but then tries to open /etc/:/hosts

**Comment by rminnich** [comment]
you might as well add this to travis.sh:
add -cmds "/etc:hosts"
and make sure it passes the bb test.

**Pull request merged**